[WPEPlatform] Support AHardwareBuffer as accelerated backing store

aperezdc opened the pull request:
#### 20d6d92cc3f91ba23ecf064dc8bb98ec9d764ec7
<pre>
[WPEPlatform] Support AHardwareBuffer as accelerated backing store
<a href="https://bugs.webkit.org/show_bug.cgi?id=294347">https://bugs.webkit.org/show_bug.cgi?id=294347</a>

Reviewed by Nikolas Zimmermann.

Enable using AHardwareBuffer-backed EGL images as render target, which
in turn allows zero-copy transport of graphics buffers with WPEPlatform
on Android.

The main changes involve teaching RenderTargetEGLImage how to create an
AHardwareBuffer-backed EGLImage, and the needed machinery to notify the
AcceleratedBackingStore about buffer creation.

Part of the changes enable seting the preferred buffer formats using the
relevant WPEPlatform API functionality, reusing the same code paths used
on regular Linux where GBM is available, converting from AHardwareBuffer
buffer format constants and back as needed to allow the public API to
continue using FourCC format codes. A default set of useable buffer
formats for the display is probed at runtime.

Canonical link: <a href="https://commits.webkit.org/304567@main">https://commits.webkit.org/304567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5969c8d83e8c6b7276b1a0871de95c8d3a66971

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136002 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8355 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143702 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/88980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137871 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9023 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/8202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/88980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138948 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/6542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/84806 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4304 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/40062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146454 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/8040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/40630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/8059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/8202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28584 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/118172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/8088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/36239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/7809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71645 "Built successfully") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/8030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/7888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,308 @@
+
+    auto* eglImage = std::exchange(priv->eglImage, EGL_NO_IMAGE);
+    auto* display = wpe_buffer_get_display(WPE_BUFFER(androidBuffer));
+    if (!display)
```
Wouldn't this leak the EGLImage, if there was no display (I know unlikely, but still)?

nikolaszimmermann replied:
Ah the callee already checks for `display` being non-null, so we can assert it here.

aperezdc replied:
I don't see `wpe_buffer_get_display()` checking that the display is a valid pointer, and the caller is `wpeBufferDispose()` and it does not check it either.

While I could move getting the `EGLImage` _after_ checking the display, the truth is that we need the display to be able to obtain the `EGLDisplay`, which is needed anyway for the `eglDestroyImageKHR()` call. No matter what we do, if the `EGLDisplay` cannot be obtained, the `EGLImage` _will_ be leaked. The `WPEBufferAndroid` instance is being destroyed here, so there's noting that would be able to keep a reference to it 😞 

The function `wpeBufferDMABufDisposeEGLImageIfNeeded()` is similar and has the same issue, FWIW.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,308 @@
+    if (!display)
+        return;
+
+    if (auto* eglDisplay = wpe_display_get_egl_display(display, nullptr)) {
```
Early return style?

nikolaszimmermann replied:
Should we warn if there is no eglDisplay? `LOG_ERROR`?

nikolaszimmermann replied:
Ok made up my mind: the callee checks for `eglDisplay` being non-null, so we can assert here and avoid the confusion (no error handling here? on purpose!)

aperezdc replied:
We can log an error, but given that the `WPEBufferAndroid` is being destroyed here, there isn't much we can do to recover from the issue.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,308 @@
+
+    if (auto* eglDisplay = wpe_display_get_egl_display(display, nullptr)) {
+        static auto s_eglDestroyImageKHR = reinterpret_cast<PFNEGLDESTROYIMAGEPROC>(epoxy_eglGetProcAddress("eglDestroyImageKHR"));
+        g_assert(s_eglDestroyImageKHR);
```
Shall we use WTF mechanisms instead? What is the preference here in WPEPlatform/wpe/ code?

aperezdc replied:
My impression was that WPEPlatform being a GObject library, it would use GObject/GLib-isms, and indeed it does in some cases: `g_debug` and `g_warning` are used in a number of places, as well as `g_assert_not_reached`, but then most other assertions use `[RELEASE_]ASSERT[_NOT_REACHED` from WTF 🤷🏼 

I suppose we should decide on one style or the other and be more consistent going forward. Probably it makes sense to use WTF for assertions, which are richer, and keep GLib for debugging as application developers are likely going to try and use e.g. `G_MESSAGES_DEBUG` with it given that WPEPlatform has a GObject API. 

@carlosgcampos Any thoughts on this?

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,308 @@
+    static PFNEGLCREATEIMAGEKHRPROC s_eglCreateImageKHR = ([eglDisplay]() {
+        if (epoxy_has_egl_extension(eglDisplay, "EGL_KHR_image_base"))
+            return reinterpret_cast<PFNEGLCREATEIMAGEKHRPROC>(epoxy_eglGetProcAddress("eglCreateImageKHR"));
+        return static_cast<PFNEGLCREATEIMAGEKHRPROC>(nullptr);
```
`return nullptr` and add `-> PFNEGLCREATEIMAGEKHRPROC` to the lambda definition.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEBufferAndroid.cpp:
```diff
@@ -0,0 +1,308 @@
+        return static_cast<PFNEGLCREATEIMAGEKHRPROC>(nullptr);
+    })();
+    if (!s_eglCreateImageKHR) {
+        g_set_error_literal(error, WPE_BUFFER_ERROR, WPE_BUFFER_ERROR_IMPORT_FAILED, "Failed to import buffer to EGL image: eglCreateImageKHR not found");
```
Should we create the s_eglCreate/Destroy static at class level, and check for both create and destroy here, instead of grabbing destroy in another place?

aperezdc replied:
I think we could grab both pointers here the first time this code path is reached. The idea of resolving the symbols here is to get the functions pointers lazily only if `wpe_buffer_import_to_egl_image()` is ever used&mdash;client code may use the buffer in some other way, which reminds me that I need to add a getter to fetch the underlying `AHardwareBuffer` from a `WPEBufferAndroid` in case one needs/wants to bypass using EGL.

nikolaszimmermann reviewed (commented):
The patch "[WPEPlatform] Introduce support for Android's AHardwareBuffer" looks good, just have a few comments. I will check the other one too.

nikolaszimmermann commented on Source/WTF/wtf/android/RefPtrAndroid.h:
```diff
@@ -0,0 +1,44 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
```
Here you used the "BY APPLE INC." - in the other commit, another non-Apple wording was chosen. I'd unify.

aperezdc replied:
Switched to the LGPL blurb, which we typically prefer for code used only by the GTK and WPE ports.

nikolaszimmermann commented on Source/WTF/wtf/android/RefPtrAndroid.h:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2025 Igalia S.L. All rights reserved.
```
Don't we usually omit the "All rights reserved."?

aperezdc replied:
Indeed. I'll change it to our usual wording.

nikolaszimmermann commented on Source/WebCore/platform/graphics/PlatformDisplay.cpp:
```diff
@@ -175,18 +175,18 @@ bool PlatformDisplay::destroyEGLImage(EGLImage image) const
 
-#if USE(GBM)
+#if USE(GBM) || OS(ANDROID)
 const Vector<GLDisplay::DMABufFormat>& PlatformDisplay::dmabufFormats()
```
Is it still accurate to talk about "dmabufFormats" - shouldn't we generalize this to "hardwareBufferFormats"? That could fit for both gbm_bo based dma-bufs, and ahbs.

aperezdc replied:
Not really, this needs renaming, the same as other parts of the code which end up being common for both AHBs and DMA-BUFs could use some cleanup and refactoring.

Question is: do we prefer to do this renaming _before_ as a preparation step, or _after_ as a follow-up? What's your take, @carlosgcampos?

aperezdc replied:
We did a few refactors/renames in `main` to this effect: https://github.com/WebKit/WebKit/pull/48535, https://github.com/WebKit/WebKit/pull/48720, and https://github.com/WebKit/WebKit/pull/48601

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
 #endif // USE(GBM)
 
+#if OS(ANDROID)
+// TODO: There is another copy of this function in AcceleratedSurfaceDMABuf.cpp, remove one.
```
Would be possible to include this fix here, no?

aperezdc replied:
There's also the AHB-format-to-FourCC conversion in `WPEBufferAndroid.cpp`, which could use an x-macro with the mapping between one and the other as well; but then we would need its definition in WTF... and it feels icky to put graphics definitions in WTF.

Alas, I am not super happy about having `WTF/wtf/android/RefPtrAndroid.{h,cpp}` in WTF, and part of me thinks that it could go in `WebCore/platform/graphics/android/RefPtrAndroid.{h,cpp}` instead.

At any rate, at least I'll move the `fourccToAHardwareBufferFormat()` function into new `WebCore/platform/graphics/android/BufferFormatAndroid.{h,cpp}` files, to be reused at both locations. Plus a comment about keeping the list in sync with the one in `WPEBufferAndroid.cpp`.

aperezdc replied:
Done, there are no longer duplicated functions, this one is now in `graphics/android/BufferFormatAndroid.{h,cpp}`.

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+    std::call_once(flag, []() {
+        formats.construct();
+
+        static constexpr uint32_t drmFormats[] = {
```
Is that a particular choice? I'd appreciate a comment about the selection of formats here.

aperezdc replied:
Yes, the choice is intentional. This list includes all those formats supported by AHB ([docs](https://developer.android.com/ndk/reference/group/a-hardware-buffer#ahardwarebuffer_format), [NDK header](https://android.googlesource.com/platform/frameworks/native/+/4f463a6b1de9198963dc6aff74154a504ba3f8f6/libs/nativewindow/include/android/hardware_buffer.h#66)) which are suitable for rendering: they have RGB color components and optionally alpha. I'll write this as a comment in the source code.

aperezdc replied:
Added a comment there to explain the choices in the list; also switched to using the `WebCore::FourCC` convenience class.

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+        };
+
+        Vector<GLDisplay::DMABufFormat> result;
+        AHardwareBuffer_Desc bufferDesc { };
```
" { }" is the default, no?

aperezdc replied:
No, because it's a local variable and we want to force zero-initialization here (omitting the `{ }` would have worked if this was a member of a class).

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+    std::call_once(flag, []() {
+        formats.construct();
+
+        static constexpr uint32_t drmFormats[] = {
```
How about a constexpr std::array?

aperezdc replied:
Well, will do.

(Although I must admit that lately I am finding `std::array` annoying because one has to specify the number of elements instead of letting the compiler figure it out by itself 🙃.)

aperezdc replied:
I learnt that C++ allows to use iteration syntax on constant plain arrays. This works:

```cpp
static constexpr FourCC drmFormats[] = { /* ... */ };

for (const auto& format : drmFormats)
    doSomething(format);
```

So I have taken this approach.

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+                static_cast<char>((drmFormat >>  0) & 0xFF),
+                '\0'
+            };
+            if (!isprint(fourccString[0])) fourccString[0] = '?';
```
This violates coding style guidelines (ok, in WPEPlatform, but still) -- how about introducing a single, multi-line formatted, helper lambda for this so we can use a self-explanatory function call here "eventuallyReplaceNonPrintableCharacter(...)"?

aperezdc replied:
Rather, now that you prompted me to search, I found we have a handy [FourCC class](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/FourCC.h). Which means we can actually use `FourCC(fourcc).string().data()` and remove code from here. Yay!

aperezdc replied:
Changed to use `FourCC`, much nicer 😄

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+
+            bufferDesc.format = fourccToAHardwareBufferFormat(drmFormat);
+            if (!bufferDesc.format) {
+                WTFLogAlways("AHB: Cannot map DRM format '%s', skipped", fourccString);
```
I'd s/AHB/AHardwareBuffer/ for clarity in the log messages.
Wouldn't LOG_ERROR fit better here? I'm often undecided myself...

aperezdc replied:
I think we may want to use `RELEASE_LOG_ERROR(IOSurface, ...)` &mdash; I am well aware that `IOSurface` is Cocoa-speak, but I think it's okay to piggyback on that log channel because it's already defined and the role of AHBs and DMA-BUFs corresponds to that of IOSurface on Cocoa.

aperezdc replied:
Changed to `RELEASE_LOG_ERROR()`.

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +235,81 @@ const Vector<GLDisplay::DMABufFormat>& GLDisplay::dmabufFormatsForVideo()
+            }
+        }
+
+        WTFLogAlways("AHB: There are %zu supported formats", result.size());
```
I also had plenty of messages like this in `MemoryMappedGPUBuffer`, that we removed prior to landing. But I agree these information are valuable, we somehow need a plan for that, e.g. some 'Info' log channel that can be easily toggled, even in release builds, to provide those traces. Probably hard to do that without runtime cost, .... I'm not sure, but wanted to mention it.

aperezdc replied:
See my suggestion above... I think `RELEASE_LOG_{ERROR,DEBUG,...}(IOSurface, ...)` would fit nicely; the logging would be disabled by default, and could be selectively enabled with `WEBKIT_DEBUG=IOSurface`.

aperezdc replied:
Done. It's nice to be able to enable those with a specific log channel.

nikolaszimmermann commented on Source/WebCore/platform/graphics/egl/GLDisplay.h:
```diff
@@ -71,6 +71,11 @@ class GLDisplay {
     const Vector<DMABufFormat>& dmabufFormatsForVideo();
 #endif
+#elif OS(ANDROID)
+    struct DMABufFormat {
```
Oh, here is the answer to a previous question -- a 'DMABufFormat' is *abused* to be something else for Android -- we should maybe just reconsider the naming scheme.

nikolaszimmermann commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -35,6 +35,10 @@
 
+#if OS(ANDROID)
+#include <wtf/android/RefPtrAndroid.h>
+#endif // OS(ANDROID)
```
This comment doesn't add much value..

nikolaszimmermann commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -50,32 +54,63 @@ static MallocSpan<uint8_t> copyBuffer(std::span<const uint8_t> buffer)
-std::unique_ptr<Decoder> Decoder::create(std::span<const uint8_t> buffer, Vector<Attachment>&& attachments)
+std::unique_ptr<Decoder> Decoder::create(std::span<const uint8_t> buffer, Vector<Attachment>&& attachments
+#if OS(ANDROID)
+        , Vector<RefPtr<AHardwareBuffer>, 2>&& hardwareBuffers
```
Why fiddle the hardwareBuffers thought static create methods, constructors etc.
Can't we just pass them after the `Decoder` creation `#if OS(ANDROID)`?

aperezdc replied:
After some talking with @carlosgcampos, I am splitting the IPC part into a separate patch, and changing the `Attachment` definition so with `OS(ANDROID)` and attachment may be either an `UnixFileDescriptor` or a `RefPtr<AHardwareBuffer>`.

nikolaszimmermann commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -226,7 +240,22 @@ bool Connection::processMessage()
+    for (size_t i = 0; i < hardwareBufferCount; ++i) {
+        AHardwareBuffer* hardwareBufferPtr { nullptr };
+        int result [[maybe_unused]] = AHardwareBuffer_recvHandleFromUnixSocket(socketDescriptor(), &hardwareBufferPtr);
+        RELEASE_ASSERT(result == 0);
```
This is guaranteed? No EXXXX codes that could be returned? (No idea about the ahb API... just guessing)

aperezdc replied:
As pointed out by @carlosgcampos in a private chat, we'll need to add error handling here, I checked the [implementation](https://android.googlesource.com/platform/frameworks/native/+/4f463a6b1de9198963dc6aff74154a504ba3f8f6/libs/nativewindow/AHardwareBuffer.cpp#392) and it only handles `EINTR` by itself. At least we'll want to retry on `-EWOULDBLOCK` given that our sockets are always configured as non-blocking.

nikolaszimmermann commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -219,6 +234,14 @@ RendererBufferFormat AcceleratedBackingStoreDMABuf::bufferFormat() const
+    else if (WPE_IS_BUFFER_ANDROID(buffer)) {
+        auto* bufferAndroid = WPE_BUFFER_ANDROID(buffer);
+        // XXX: Given that we transform the AHardwareBuffer format into a DRM fourcc, we can abuse ::Type::DMABuf here.
+        format.type = RendererBufferFormat::Type::DMABuf;
```
I'd name it `HardwareBuffer` instead in a different pre-requisite commit.

aperezdc replied:
I've added a new entry to the `enum`, this is used only to describe the buffers being used in the `webkit://gpu` page and it seems worth it to be precise with what we show there.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+#endif // USE(GBM)
+
+#if OS(ANDROID)
+// TODO: There is another copy of this function in GLDisplay.cpp, remove one.
```
A common `AHardwareBufferUtilities.h`?

aperezdc replied:
Yes, as I outlined in a previous comment.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+usageToAHardwareBufferUsage(DMABufRendererBufferFormat::Usage usage)
+{
+    switch (usage) {
+    case DMABufRendererBufferFormat::Usage::Rendering:
```
Does that need `AHARDWAREBUFFER_USAGE_CPU_READ_RARELY` ? Do we loose the ability to read-back when omitting it?

aperezdc replied:
The documentation says:

> The buffer will sometimes be locked for direct CPU reads using the [AHardwareBuffer_lock()](https://developer.android.com/ndk/reference/group/a-hardware-buffer#group___a_hardware_buffer_1gaf10c050f1ddfe2a074ab0f80660b2925) function.
>
> Note that reading the buffer using OpenGL or Vulkan functions or memory mappings does not require the presence of this flag.

At the moment we are not using that anywhere. I was planning to use `AHardwareBuffer_lock()` when implementing the `WPEBuffer.import_to_pixels` vfunc later (the alternative being circuitous: creating an `EGLImage`, binding it as a texture, setting it as a FBO, and then `glReadPixels`).

I will leave the flag out for now, and we can reconsider it when/if we add an `.import_to_pixels` implementation for `WPEBufferAndroid`.

aperezdc replied:
I've filed [an issue](https://bugs.webkit.org/show_bug.cgi?id=297316) to implement `.import_to_pixels` later on, added a `FIXME` referencing it, and edited the flag selection.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+        return nullptr;
+    }
+
+    AHardwareBuffer_Desc description = { };
```
s/ { } /

aperezdc replied:
Does C++ zero-initialize structs defined as plain C ones if we leave out the explicit `{ }` brackets? Because we definitely want to zero-initialize here. I'll double check.

aperezdc replied:
I checked this, and as I was expecting we need either the trailing `= { }` or `{ }`  to tell the compiler to perform zero-initialization.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+    AHardwareBuffer* hardwareBufferPtr { nullptr };
+    int result = AHardwareBuffer_allocate(&description, &hardwareBufferPtr);
+    if (result) {
+        WTFLogAlways("Failed to create Android buffer of size %dx%d, format %#0" PRIx32 ", error code: %d",
```
LOG_ERROR?

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+    }
+    RefPtr hardwareBuffer = adoptRef(hardwareBufferPtr);
+
+    const Vector<EGLAttrib> attributes { EGL_IMAGE_PRESERVED, EGL_TRUE, EGL_NONE };
```
You used std::array in another place.

aperezdc replied:
...but not here because `GLDisplay::createImage()` takes a `Vector<EGLAttrib>` parameter.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -277,7 +283,95 @@ AcceleratedSurfaceDMABuf::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t su
+    auto clientBuffer = eglGetNativeClientBufferANDROID(hardwareBuffer.get());
+    auto image = display.createEGLImage(EGL_NO_CONTEXT, EGL_NATIVE_BUFFER_ANDROID, clientBuffer, attributes);
+    if (image == EGL_NO_IMAGE) {
+        WTFLogAlways("Failed to bind Android buffer to an EGLImage. This is typically caused by "
```
Ditto.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
```diff
@@ -41,10 +41,17 @@
 
+#if OS(ANDROID)
+typedef struct AHardwareBuffer AHardwareBuffer;
+#endif // OS(ANDROID)
```
The other #endif don't have the trailing comment, I think you can omit it.

aperezdc replied:
I think for the final version of the patch I'll remove the trailing `// OS(ANDROID)` in those cases where the corresponding `#if` is close by and it's obvious what the `#endif` is closing. Other ones, where the preprocessor directives are far from each other, are quite useful.

nikolaszimmermann commented on Source/cmake/FindAndroid.cmake:
```diff
@@ -87,6 +87,7 @@ function(_AndroidHandleComponent target)
 
     check_include_file("${header}" Android_COMPONENT_${target}_HAS_HEADER)
     if (NOT Android_COMPONENT_${target}_HAS_HEADER)
+        message(STATUS "missing header ${header}")
```
s/mis/Mis/?

nikolaszimmermann reviewed (commented):
Phew, a lot of stuff :-) Thanks @aperezdc - looks mostly clean.

carlosgcampos commented on Source/WebCore/platform/graphics/PlatformDisplay.cpp:
```diff
@@ -173,18 +173,18 @@ bool PlatformDisplay::destroyEGLImage(EGLImage image) const
 
-#if USE(GBM)
+#if USE(GBM) || OS(ANDROID)
 const Vector<GLDisplay::DMABufFormat>& PlatformDisplay::dmabufFormats()
```
I guess we should rename GLDisplay::DMABufFormat too at some point then

aperezdc replied:
I'm working on the rename, it's tracked in [issue #298309](https://bugs.webkit.org/show_bug.cgi?id=298309).

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -316,7 +323,74 @@ AcceleratedSurface::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t surfaceI
+
+#if OS(ANDROID)
+static uint64_t
+usageToAHardwareBufferUsage(RendererBufferFormat::Usage usage)
```
```suggestion
static uint64_t usageToAHardwareBufferUsage(RendererBufferFormat::Usage usage)
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -316,7 +323,74 @@ AcceleratedSurface::RenderTargetEGLImage::RenderTargetEGLImage(uint64_t surfaceI
+    case RendererBufferFormat::Usage::Mapping:
+        return AHARDWAREBUFFER_USAGE_CPU_READ_RARELY | AHARDWAREBUFFER_USAGE_GPU_FRAMEBUFFER;
+    case RendererBufferFormat::Usage::Scanout:
+        return AHARDWAREBUFFER_USAGE_CPU_READ_RARELY | AHARDWAREBUFFER_USAGE_GPU_FRAMEBUFFER | AHARDWAREBUFFER_USAGE_FRONT_BUFFER;
```
Is this really equivalent to scanout?

aperezdc replied:
I think so, the [commend in the header](https://android.googlesource.com/platform/frameworks/native/+/4f463a6b1de9198963dc6aff74154a504ba3f8f6/libs/nativewindow/include/android/hardware_buffer.h#267) reads:

```cpp
     /** The buffer will be written to by the GPU as a framebuffer attachment.*/
```

I do have a change pending to push here, though! When integrating WPEPlatform into the WPE-Android widgetry, we will want to have the `AHARDWAREBUFFER_USAGE_COMPOSER_OVERLAY` flag added as well:

```cpp
   /**
     * The buffer will be used as a composer HAL overlay layer.
     *
     * This flag is currently only needed when using ASurfaceTransaction_setBuffer
     * to set a buffer. In all other cases, the framework adds this flag
     * internally to buffers that could be presented in a composer overlay.
     * ASurfaceTransaction_setBuffer is special because it uses buffers allocated
     * directly through AHardwareBuffer_allocate instead of buffers allocated
     * by the framework.
     */
    AHARDWAREBUFFER_USAGE_COMPOSER_OVERLAY      = 1ULL << 11,
```

...because what we _precisely_ do there is using `ASurfaceTransaction` to configure a given buffer as the contents of the widget, and with this patch we switch to allocating the buffers ourselves instead of getting them from the Android frameworks&mdash;as stated in the documentation comment.

aperezdc replied:
Now, to round up things, even if `AHARDWAREBUFFER_USAGE_FRONT_BUFFER` might look like it's unneeded when using `AHARDWAREBUFFER_USAGE_COMPOSER_OVERLAY`, the [Java documentation](https://android.googlesource.com/platform/frameworks/base/+/1cdfff555f4a21f71ccc978290e2e212e2f8b168/core/java/android/hardware/HardwareBuffer.java#201) is more complete:

```java
/** Usage: The buffer is used for front-buffer rendering. When front-buffering rendering is
 * specified, different usages may adjust their behavior as a result. For example, when
 * used as USAGE_GPU_COLOR_OUTPUT the buffer will behave similar to a single-buffered window.
 * When used with USAGE_COMPOSER_OVERLAY, the system will try to prioritize the buffer
 * receiving an overlay plane & avoid caching it in intermediate composition buffers. */
public static final long USAGE_FRONT_BUFFER           = 1L << 32;
```

In particular:

> When used with `USAGE_COMPOSER_OVERLAY`, the system will try to prioritize the buffer receiving an overlay plane & avoid caching it in intermediate composition buffers.

spenap commented on Source/WebCore/platform/graphics/PlatformDisplay.cpp:
```diff
@@ -173,18 +173,18 @@ bool PlatformDisplay::destroyEGLImage(EGLImage image) const
+#endif // USE(GBM) || OS(ANDROID)
 
-#if USE(GSTREAMER)
+#if USE(GBM) && USE(GSTREAMER)
```
This requires both `USE(GBM) && USE(GSTREAMER)` for `dmabufFormatsForVideo`. In the header, it could be declared if `OS(ANDROID) && USE(GSTREAMER)` even if `USE(GBM)` is false

aperezdc replied:
That is intentional, this patch does *not* try to probe and provide a list of buffer formats suitable for video content. The idea is to focus here on web page composition, further optimizations like avoiding pixel format conversions for video can be done later.

spenap commented on Source/WebCore/platform/graphics/egl/GLDisplay.h:
```diff
@@ -70,6 +70,11 @@ class GLDisplay final : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<G
+    struct DMABufFormat {
+        uint32_t fourcc;
+    };
+    const Vector<DMABufFormat>& dmabufFormats();
```
I guess this could be `const` if we consider the static logic a mutable cache that is only set once

aperezdc replied:
You mean the `BufferFormat::fourcc` member? (note that `DMABufFormat` was renamed as `BufferFormat` recently). I suppose it could work, but I would rather do that in a follow-up patch.

carlosgcampos commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +238,45 @@ const Vector<GLDisplay::BufferFormat>& GLDisplay::bufferFormatsForVideo()
+            RELEASE_ASSERT(ahbFormat);
+            bufferDesc.format = ahbFormat.value();
+            if (AHardwareBuffer_isSupported(&bufferDesc)) {
+                RELEASE_LOG_DEBUG(IOSurface, "AHB: Adding supported DRM format '%s'", drmFormat.string().data());
```
IOSurface?

aperezdc replied:
It seemed fitting: the logging channel was already defined, and `AHardwareBuffer` on Android is the equivalent to `IOSurface` on MacOS/iOS 😉 

We can rename the logging channel to something like `GraphicsBuffer` and have all the port use that. Or we can put the `IOSurface` log channel definition under `#if PLATFORM(COCOA)` and add a new one for our ports. But IMO that can be done separately.

carlosgcampos replied:
It's confusing, I think there are more generic channels like Graphics that we can use, no?

aperezdc replied:
There is no `Graphics` channel in WebCore. I ended up adding a new `GraphicsBuffer` channel, and later I can try to make a follow-up patch changing logging statements which use `IOSurface` to use `GraphicsBuffer` (and remove `IOSurface`).

carlosgcampos commented on Source/WebCore/platform/graphics/egl/GLDisplay.h:
```diff
@@ -67,16 +67,19 @@ class GLDisplay final : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<G
+#endif
     };
     const Vector<BufferFormat>& bufferFormats();
+#endif // USE
```
```suggestion
#endif
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -182,7 +187,7 @@ Vector<RendererBufferFormat> WebPageProxy::preferredBufferFormats() const
 }
 
-#if USE(GBM) && ENABLE(WPE_PLATFORM)
+#if ENABLE(WPE_PLATFORM)
```
```suggestion
#if ENABLE(WPE_PLATFORM) && (USE(GBM) || OS(ANDROID))
```

carlosgcampos replied:
But I don't think this can be called in android, so I would keep it GBM only.

aperezdc replied:
The suggested change is not needed because this part is already inside an `USE(GBM) || OS(ANDROID)` guard starting in line 126, something like this:

```cpp
#if USE(GBM) || OS(ANDROID)

// Long piece of code, taller than a screenful. 

#if ENABLE(WPE_PLATFORM)
void WebPageProxy::preferredBufferFormatsDidChange()
{
    // ...
}
#endif
#endif
```

This is a clear case where annotating the `#endif` with a comment that contains corresponding guard condition makes it easier to reason about the code. I am going to annotate the second `#endif` there to be `#endif // USE(GBM) || OS(ANDROID)` 😉

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -623,16 +703,14 @@ bool AcceleratedSurface::SwapChain::resize(const IntSize& size)
     case Type::EGLImage:
         return RenderTargetEGLImage::create(m_surfaceID, m_size, m_bufferFormat);
 #endif
     case Type::Texture:
         return RenderTargetTexture::create(m_surfaceID, m_size);
     case Type::SharedMemory:
         return RenderTargetSHMImage::create(m_surfaceID, m_size);
```
This should be #if PLATFORM(GTK) || ENABLE(WPE_PLATFORM)

aperezdc replied:
Do you mean that we should not provide a SHM fallback on Android at all?

carlosgcampos replied:
No, I mean those targets should only be available for gtk and wpe platform, but not in WPE when building without wpe platform.

aperezdc replied:
Got it, guards rearranged.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -735,7 +813,7 @@ void AcceleratedSurface::SwapChain::addDamage(const std::optional<Damage>& damag
 }
 #endif
 
-#if PLATFORM(WPE) && USE(GBM) && ENABLE(WPE_PLATFORM)
+#if PLATFORM(WPE) && ENABLE(WPE_PLATFORM) && (USE(GBM) || OS(ANDROID))
```
I don't think this can happen in android

aperezdc replied:
It could happen if the `WPEView::preferred-dma-buf-formats-changed` signal gets emitted, which the platform implementation for Android might want to do: it's possible that the pixel format used for a native window gets changed (e.g. there is an `ANativeWindow_setWindowFormat` native function, and I think an equivalent Java one), and even when I am not completely sure if this will be done by the system frameworks, if the window format is changed for any reason the platform implementation may want to notify about this through WPEPlatform to make WebKit change buffers of views to be displayed in the corresponding native window to follow.

So actually what is missing is is enabling handling the `WPEView::preferred-dma-buf-formats-changed` signal in `WPEWebViewPlatform.cpp`.

aperezdc replied:
Actually, shouldn't we rename `WPEView::preferred-dma-buf-formats-changed` to something like `WPEView::preferred-buffer-formats-changed`, `wpe_view_get_preferred_dma_buf_formats()` to `wpe_view_get_preferred_buffer_formats()`, and so on?

carlosgcampos replied:
Exactly, I meant it was not possible because WPEWebViewPlatform only monitors the signal for GBM. And yes, I think we should rename the api to make it dmabuf-less.

aperezdc replied:
The rename was done in https://github.com/WebKit/WebKit/pull/50232

Note that the patch changes the guard to `USE(GBM) || OS(ANDROID)` when connecting to the signal to monitor the preferred buffer formats.

spenap commented on Source/WebCore/platform/graphics/egl/GLDisplay.h:
```diff
@@ -67,16 +67,19 @@ class GLDisplay final : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<G
     const Vector<BufferFormat>& bufferFormats();
+#endif
+
 #if USE(GSTREAMER)
```
should this now be `#if USE(GBM) && USE(GSTREAMER)` ?

aperezdc replied:
Indeed, changed.

spenap commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -135,8 +135,11 @@ Vector<RendererBufferFormat> WebPageProxy::preferredBufferFormats() const
     if (!formats)
         return { };
 
+#if USE(GBM)
+    WPEDRMDevice* mainDevice = wpe_buffer_buffer_formats_get_device(formats);
+#endif
+
```
I would remove this from here...

spenap commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -152,11 +155,13 @@ Vector<RendererBufferFormat> WebPageProxy::preferredBufferFormats() const
+#if USE(GBM)
         WPEDRMDevice* targetDevice = wpe_buffer_formats_get_group_device(formats, i);
         if (!targetDevice)
             targetDevice = mainDevice;
```
```suggestion
            targetDevice = wpe_buffer_buffer_formats_get_device(formats);
```

and add it here. Unless the diff confused me, it doesn't seem like `mainDevice` is used elsewhere, and `formats` is sed in a const-way (although that's not explicit. Should we?)

This allows us to remove the guards earlier.

aperezdc replied:
The idea is to use the `wpe_buffer_formats_get_device()` getter only once before the loop... although, given that it is a simple getter, maybe it's fine to move the call around as you suggest.

spenap replied:
Sorry! I totally missed this was in a loop: in that case, I have no strong opinion :pray:

aperezdc replied:
I left it outside of the loop in the end.

spenap commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2862,7 +2863,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void setToolbarsAreVisible(bool);
 
-#if USE(GBM) && (PLATFORM(GTK) || PLATFORM(WPE))
+#if (USE(GBM) || OS(ANDROID)) && (PLATFORM(GTK) || PLATFORM(WPE))
```
Totally minor, but would it make sense to use the same order consistently to facilitate future searches?


```suggestion
#if (PLATFORM(GTK) || PLATFORM(WPE)) && (USE(GBM) || OS(ANDROID))
```

Like line 619

spenap reviewed (commented):
Very nice one, what a best!

felipeerias commented on Source/WebCore/platform/graphics/egl/GLDisplay.cpp:
```diff
@@ -231,4 +238,45 @@ const Vector<GLDisplay::BufferFormat>& GLDisplay::bufferFormatsForVideo()
+            const auto ahbFormat = toAHardwareBufferFormat(drmFormat);
+            RELEASE_ASSERT(ahbFormat);
+            description.format = ahbFormat.value();
+            if (AHardwareBuffer_isSupported(&description)) {
```
This may return false because the buffer description is not valid, because its size and its number of layers are zero.

From `AHardwareBuffer_allocate()` in [AHardwareBuffer.cpp](https://android.googlesource.com/platform/frameworks/native/+/refs/heads/main/libs/nativewindow/AHardwareBuffer.cpp?pli=1#140):

```cpp
int AHardwareBuffer_allocate(const AHardwareBuffer_Desc* desc, AHardwareBuffer** outBuffer) {
    if (!outBuffer || !desc) return BAD_VALUE;
    if (!AHardwareBuffer_isValidDescription(desc, /*log=*/true)) return BAD_VALUE;
```

And then `AHardwareBuffer_isValidDescription()`:

```cpp
bool AHardwareBuffer_isValidDescription(const AHardwareBuffer_Desc* desc, bool log) {
    if (desc->width == 0 || desc->height == 0 || desc->layers == 0) {
        ALOGE_IF(log, "Width, height and layers must all be nonzero");
        return false;
    }
```
---

I am not sure if this naive fix would be the best solution, but it seems to work in my tests (using Android 16):

```cpp
    AHardwareBuffer_Desc description { };
    description.width = 1;
    description.height = 1;
    description.layers = 1;
    description.usage = AHARDWAREBUFFER_USAGE_GPU_FRAMEBUFFER | AHARDWAREBUFFER_USAGE_GPU_SAMPLED_IMAGE;
```

aperezdc replied:
I actually had something like this at some point, but then removed it because the [documentation](https://developer.android.com/ndk/reference/group/a-hardware-buffer#ahardwarebuffer_issupported) reads that all fields except the stride may be taken into account, but there is no note telling that they can't be zero 🙃 

Fixed.

nikolaszimmermann commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
```diff
@@ -145,6 +147,17 @@ void AcceleratedBackingStore::didCreateSHMBuffer(uint64_t id, WebCore::Shareable
+{
+    RELEASE_ASSERT(hardwareBuffer);
+
+    GRefPtr buffer = adoptGRef(WPE_BUFFER(wpe_buffer_android_new(wpe_view_get_display(m_wpeView.get()), hardwareBuffer.get())));
```
auto?

nikolaszimmermann commented on Source/WebKit/UIProcess/glib/RendererBufferDescription.h:
```diff
@@ -31,7 +31,11 @@
+    enum class Type : uint8_t {
+        DMABuf,
+        SharedMemory,
+        AHardwareBuffer,
```
Shouldn't this be OS(ANDROID) guarded?

aperezdc replied:
Could be, I was trying to avoid unneeded guards, but there is not much of a point of having all the rest with `OS(ANDROID)` and not this small bit. I'll add it before landing.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -306,15 +313,90 @@ std::unique_ptr<AcceleratedSurface::RenderTarget> AcceleratedSurface::RenderTarg
+            size.width(), size.height(), FourCC(bufferFormat.fourcc).string().data(), result);
+        return nullptr;
+    }
+    RefPtr hardwareBuffer = adoptRef(hardwareBufferPtr);
```
auto?

nikolaszimmermann reviewed (approved):
Looks great, let's get this in finally :-)

aperezdc committed:
[WPEPlatform] Support AHardwareBuffer as accelerated backing store
https://bugs.webkit.org/show_bug.cgi?id=294347

Reviewed by Nikolas Zimmermann.

Enable using AHardwareBuffer-backed EGL images as render target, which
in turn allows zero-copy transport of graphics buffers with WPEPlatform
on Android.

The main changes involve teaching RenderTargetEGLImage how to create an
AHardwareBuffer-backed EGLImage, and the needed machinery to notify the
AcceleratedBackingStore about buffer creation.

Part of the changes enable seting the preferred buffer formats using the
relevant WPEPlatform API functionality, reusing the same code paths used
on regular Linux where GBM is available, converting from AHardwareBuffer
buffer format constants and back as needed to allow the public API to
continue using FourCC format codes. A default set of useable buffer
formats for the display is probed at runtime.

Canonical link: https://commits.webkit.org/304567@main
